cli/data.go

```diff
@@ -0,0 +1,194 @@
+		if err != nil {
+			return err
+		}
+
+		// TODO: We need to store file extension too. In sync we map from ext -> mime type, so this is already available.
+		//       Or maybe we can just get this from file name.
+		ext := mimeTypeToFileExt(md.GetMimeType())
+
+		dataFile, err := os.Create(filepath.Join(dst, "data", datum.GetId()+ext))
+		if _, err := io.Copy(dataFile, r); err != nil {
+			return err
+		}
+		if err := r.Close(); err != nil {
+			return err
+		}
+		skip++
+	}
+
+	return nil
+}
+
+func (c *AppClient) TabularData(dst string, filter *datapb.Filter) error {
+	if err := c.ensureLoggedIn(); err != nil {
+		return err
+	}
+
+	// Make dst/data and dst/metadata directory.
+	if err := os.MkdirAll(filepath.Join(dst, dataDir), os.ModePerm); err != nil {
+		return err
+	}
+	if err := os.MkdirAll(filepath.Join(dst, metadataDir), os.ModePerm); err != nil {
+		return err
+	}
+
+	resp, err := c.dataClient.TabularDataByFilter(context.Background(), &datapb.TabularDataByFilterRequest{
+		DataRequest: &datapb.DataRequest{
+			Filter: filter,
+			// TODO: For now don't worry about skip/limit. Just do everything in one request. Can implement batching when
+			//       tabular is implemented.
+		},
+		CountOnly: false,
+	})
+	if err != nil {
+		return err
+	}
+	mds := resp.GetMetadata()
+
+	for i, md := range mds {
+		mdJsonBytes, err := protojson.Marshal(md)
+		if err != nil {
+			return errors.Wrap(err, "error marshaling metadata")
+		}
+		mdFile, err := os.Create(filepath.Join(dst, "metadata", strconv.Itoa(i)+".json"))
+		if err != nil {
+			return errors.Wrapf(err, "error creating metadata file %s", mdFile.Name())
+		}
+		if _, err := mdFile.Write(mdJsonBytes); err != nil {
+			return errors.Wrapf(err, "error writing metadata file %s", mdFile.Name())
+		}
+		if err := mdFile.Close(); err != nil {
+			return errors.Wrapf(err, "error closing metadata file %s", mdFile.Name())
+		}
+	}
+
+	data := resp.GetData()
+	// TODO: Use textpb insted of ndjson, and save multiple files.
+	dataFile, err := os.Create(filepath.Join(dst, "data", "data"+".ndjson"))
+	w := bufio.NewWriter(dataFile)
+	for _, datum := range data {
+		// Write everything as json for now.
+		d := datum.GetData()
+		if d == nil {
+			// TODO: This should never happen. Should this notify user? Error? Or just skip?
+			fmt.Println("Received empty tabular data")
+			continue
+		}
+		m := d.AsMap()
+		m["TimeRequested"] = datum.GetTimeRequested()
+		m["TimeReceived"] = datum.GetTimeReceived()
+		j, err := json.Marshal(m)
+		if err != nil {
+			return errors.Wrap(err, "error marshaling json response")
+		}
+		_, err = w.Write(append(j, []byte("\n")...))
+		if err != nil {
+			return errors.Wrapf(err, "error writing reading to file %s", dataFile.Name())
+		}
+	}
+
+	if err := w.Flush(); err != nil {
+		return errors.Wrapf(err, "error flushing writer for %s", dataFile.Name())
+	}
+
+	return nil
+}
+
+func mimeTypeToFileExt(mime string) string {
```

Comment:
+1 to your current work to include this in the upload/storage/response. Good example of hackier fixes getting exponentially more untenable, as we've repeated this pattern too frequently throughout the codebase.